src/constfold.c

```diff
@@ -901,6 +901,11 @@ Expression *Equal(enum TOK op, Type *type, Expression *e1, Expression *e2)
                 if (cmp == 0)
                     break;
             }
+            if (cmp && es1->type->needsNested())
+            {
+                if ((es1->sinit != NULL) != (es2->sinit != NULL))
```

Comment:
When I wrote this code, I was thinking that setting `sinst` would have been represent that the hidden field (`vthis`) of StructLiteralExp is set to null at runtime.

But the assumption was not correct, because `sinit` setting was not enough for many other bugs around nested structs. Today, as far as I remember, the null setting of runtime hidden field is represented by `StructLiteralExp.elements.dim == sd.field.dim`, and the unsetting of hidden field (== it's set to the enclosing scope frame pointer at runtime) is represented by `StructLiteralExp.elements.dim == sd.field.dim - 1`. (Note that CTFE is using a bit different representation to distinguish the two. If you interest, you can read `StructLiteralExp::interpret()`.)

So today, the `sinit` comparison would be made unnecessary if I'm not mistaken.